### ISSUE TITLE
Bug on widgets.php

### ISSUE THREAD

**Issue by itibook** [easydigitaldownloads/Easy-Digital-Downloads#3177]
Hello,

I spent a few hours on this one, unsure if this is the wanted behaviour or not, but here's my findings...

My widgets.php was breaking right after the Download Detail widget. As soon as I unregistered the widget everything went back to normal. I also had issue on the customise.php that was not working.

In /includes/widgets.php at line 302 there's the args array for the get_posts function. 

		$args = array( 
	            'post_type'      => 'download', 
	            'posts_per_page' => -1, 
	            'post_status'    => 'publish', 
	        );
$downloads = get_posts( $args );

I am not sure if the -1 for the post_per_page is the meant to be that way, but in my case, where I have thousands of downloads it seems that it had a negative effect (sometimes also had memory issues popping up as errors).

as soon as I changed the -1 to a normal number (e.g. 20) the issue disappeared.

could you please confirm if this is an issue only affecting installation with large amount of downloads or if the -1 is indeed a bad practice and should be changed to something more sensible. I am not sure if there are other implications by changing the -1 (I did not see any, but who knows..)

ciao

Luca

**Comment by danieliser** [comment]
Hey Luca, Your correct in the assumption its bad for sites with lots of downloads. The alternative unfortunately will require quite a bit of work to set up and would be an ajax search input similar to how tags and discounts work.

**Comment by itibook** [comment]
Hi Daniel,

I am not really clear how the widget works and how it impacts other areas. any idea where changing the -1  to let's say 20 could have a negative impact?

From what I see I only use it in the sidebar for a single download, so I am not quite clear why it has to pick up all the other downloads in the dropdown menu, but maybe I am missing the point. Basically I don't see the functional need for that.

What about creating two version of widget? one that only uses the current download and the other one that has the full drop down menu...

or there could be an setting in EDD to limit the amount posts (i.e. the -1 parameter) to show in the dropdown menu

any suggestion on how I could best work around this issue will be appreciated. Not really a developer, so any hint/tip is gold :-)

ciao

Luca

**Comment by mindctrl** [comment]
Luca,

If we change the -1 to 20 as you suggest, you would only see 20 items in the dropdown when adding the widget to a sidebar area, leaving you no way to select the vast majority of your downloads. As Daniel noted, to better accommodate sites with a lot of downloads, that section probably needs to be reworked to allow for searching from the dropdown.

**Comment by danieliser** [comment]
Correct, limiting it to 20 if you say have 100 downloads, you will only ever be able to choose the first 20 downloads.

The best method would be to setup a text input / searchable dropdown that uses ajax to load beyond a certain count.

**Comment by itibook** [comment]
yes, I noticed the effect of changing to 20... for me it doesn't have an impact as I only use the first option (i.e. Use current).

Thinking out loud here, to avoid going into text input, a quick option could be to do an ajax call to populate the dropdown list only if the user clicks on the list to change the Use Current value.

Currently my concern is how can I keep using EDD with a workable solution. Cause going into widgets.php and change the -1 everytime there's an update is not a long term solution.

Any way I could work around this through the child theme? via functions.php?

ciao

Luca

**Comment by pippinsplugins** [comment]
We should really change this widget to use the product drop down in the HTML elements class. That displays limited results and uses ajax to search for additional products.

**Comment by danieliser** [comment]
Bingo, there you go. I for one didn't even realize that the HTML class had this built in already. Makes the changes even easier. If i find time this week I will look at what needs to change and see about getting a PR submitted.

**Comment by pippinsplugins** [comment]
Milestoned for 2.4.

**Comment by danieliser** [comment]
Go ahead and assign this one to me. Makes it easier for me to find the issues I'm working on.

**Comment by itibook** [comment]
impressed with the response guys... :-)

thanks for looking into it, really appreciated!

**Comment by danieliser** [comment]
@pippinsplugins One problem Im seeing here with using the HTML Elements class here is the inability to inject an option for 'current'. Easy fixes seem to be setting 'show_option_none' to true and changing the usage during render of the widget to use an empty value in place of current. Essentially replacing current with none.

But that won't be self explanatory for users in my oppinion. Would adding a new filter to the HTML Elements product_dropdown options be viable? Would allow passing an extra option with the array.

**Comment by danieliser** [comment]
Also need to now load the EDD JS on widgets.php for the AJAX chosen select to function. Thought on the best way to do so? Use filter on 'edd_load_admin_scripts', modify edd_is_admin_page() etc.

**Comment by danieliser** [comment]
To get this working this is the changes I've had to make so far.

One line in edd-admin.css to override chosen select width inside widgets.
Added widgets.php to edd_is_admin_page().
Modified the widget form.

The problem though is that new widgets when created don't instantiate the chosen.js until the page is refreshed. So I will have to add some JS to initialize those new widgets on creation, no biggie but can't get around it.

**Comment by pippinsplugins** [comment]
`selected` is what you want to use for telling EDD what the current option is.

**Comment by danieliser** [comment]
@pippinsplugins - Sorry let me clarify. I mean in the widget dropdown currently the first option available is "Currentt" or the current download rather than a specific ID. Meaning it shows the download info for whatever download page it is on. Using the product_dropdown, I can't inject the extra <option value="current"><?php _e( 'Current Download', 'edd' ); ?></option> that is currently there.

**Comment by pippinsplugins** [comment]
Ah, I see what you mean.

What if we used two drop downs? One to choose `Current` or `Specific`. If they choose `Specific`, we then show a new download that has the options.

**Comment by danieliser** [comment]
I can do that for sure if that's your preferred solution. Filtering the select may be an alternative but is only useful if there are other potential uses.

**Comment by danieliser** [comment]
Also what would your suggestion be to migrate existing widgets without breakage if we create a second dropdown? Simply do a detection of a proper ID or current and then just fallback properly I'm guessing?

**Comment by cklosowski** [comment]
As far as the JS is concerned, it'd be best to break the JS for this dropdown into the front end JS file, not load the admin js on the front end as well. The other option is to just include it in both. What's your current plan @danieliser on where to put the JS for the HTML products dropdown?

**Comment by jasongill** [comment]
There is a lot of discussion here and in #4217 about making something work with Javascript (which seems like overkill, IMHO, since this is a really rare edge case). Since most sites that have tens of thousands of downloads aren't likely to be adding widgets to show details of a specific product, and are more likely to be maintained by someone who has a bit of experience, perhaps there is a better interim solution.

What if, in the event that there are more than ~1000 products (or some other "big number), the widget form simply changes to have a radio button for "current" or "specific", with the specific choice having an text input field to enter a download by ID (or slug)?

**Comment by itibook** [comment]
agree with Jason...

**Comment by felixarntz** [comment]
@jasongill:
This would work for me as well. Creating an autocomplete with JS would still be more user-friendly, but I also see how it's overkill for this edge-case - and an experienced webmaster should be able to work with your proposal like you said.

**Comment by pippinsplugins** [comment]
I'm happy with that solution.

@cklosowski Let's make it happen for 2.5.10 please.

**Comment by cklosowski** [comment]
So a radio for `current` or `specific download`, that exposes a text field to enter in a download ID? Just making sure I'm reading this right.

**Comment by pippinsplugins** [comment]
Yep, that works for me.

**Comment by cklosowski** [comment]
:+1:

**Comment by cklosowski** [comment]
How about this...for when there are less than 1000 products

![screen shot 2016-03-09 at 4 36 38 pm](https://cloud.githubusercontent.com/assets/1390124/13654385/45ee55fc-e615-11e5-96ce-f3b674b9c0e8.png)

When it's 1000 or more, it changes to a text input where you put in a download ID
![screen shot 2016-03-09 at 4 39 02 pm](https://cloud.githubusercontent.com/assets/1390124/13654425/77636d2a-e615-11e5-852c-acaa321a093a.png)

**Comment by SDavisMedia** [comment]
@cklosowski Test working fine for me! :+1: Great idea, @jasongill.

**Comment by jasongill** [comment]
This is absolutely perfect - thanks @cklosowski for delivering and coming up with a fix for an issue that drove me (and it sounds like @itibook too) crazy trying to figure out :)

:+1: :+1:

**Comment by itibook** [comment]
happy to see this being solved! And with a very pragmatic solution too!

**Comment by pippinsplugins** [comment]
Excellent.

I'd like to see the input field label say `Download ID` instead of just `Download` to ensure there is no confusion as to what should be entered.

**Comment by cklosowski** [comment]
The placeholder says Doenload ID otherwise it's just swapping the form
elements. Not the wrapper.
-- 
Chris Klosowski

**Comment by pippinsplugins** [comment]
Ah, I hadn't seen it with the placeholder. I'm good with this then.

**Comment by pippinsplugins** [comment]
Looks like a unit test needs updated.

**Comment by cklosowski** [comment]
@pippinsplugins Unit tests updated.
